libcxx/include/support/atomic/cxx_atomic.h

```diff
@@ -14,8 +14,17 @@
 template <typename _Tp, int _Sco>
 struct __cxx_atomic_base_impl {
   using __underlying_t = _Tp;
+  using __load_t = __cxx_atomic_base_impl<_Tp, _Sco>;
+  using __temporary_t = __cxx_atomic_base_impl<_Tp, _Sco>;
+  using __wrap_t = __cxx_atomic_base_impl<_Tp, _Sco>;
```

Comment:
What are these for?

Reply:
`__load_t` is unused, deleting it.
`__temporary/wrap_t` are part of some machinery for loading either the wrapping structure or the raw pointer into intrinsics in `atomic_base.h`. 

I think it's kind of a mess, but I wanted to deduplicate code paths for `volatile`, `const`, and add the new __cxx_atomic_ref backend.

Reply:
```
template <typename _Tp, typename _Up>
inline auto __cxx_atomic_exchange(_Tp* __a, _Up __val,
                          memory_order __order) -> __cxx_atomic_underlying_t<_Tp> {
  auto __v_temp = __cxx_atomic_wrap_to_base(__a, __val);
  auto __ret = __cxx_atomic_base_temporary(__a);
  __atomic_exchange(__cxx_atomic_unwrap(__a), &__v_temp, &__ret, __cxx_atomic_order_to_int(__order));
  return *__cxx_get_underlying_atomic(&__ret);
}
```

For example, this needs to exchange `__cxx_atomic_base_impl<_Tp>` so it wraps the given value into the same alignment using `__cxx_atomic_wrap_to_base` and creates a temporary to return to. 

If this was an `atomic_ref` we instead unwrap completely and pass the pointer and given user parameter directly.